### ISSUE TITLE
No type narrowing for discriminated union type guards on unique symbol properties

### ISSUE THREAD

**Issue by kpdonn** [Microsoft/TypeScript#23135]
***Note: I know about #10530 but I explain why I think this deserves its own issue in the related issues section below***

**TypeScript Version:**  2.9.0-dev.20180403

**Search Terms:** 
unique symbol, discriminated union, type guard, narrow, narrowing

**Code**

```ts
declare const kind: unique symbol

type Square = {
  readonly [kind]: "square"
  length: number
  width: number
}

type Circle = {
  readonly [kind]: "circle"
  radius: number
}

type Shape = Square | Circle

function doesNotNarrow(shape: Shape) {
  if (shape[kind] === "square") {
    console.log(shape.length, shape.width) // error length does not exist on Shape 
  } else {
    console.log(shape.radius) // error radius does not exist on Shape 
  }
}
```
**Expected behavior:**
No errors because the type should be narrowed inside the if statements.

**Actual behavior:**
Compile errors because typescript does not narrow the type despite the discriminated union.

**Playground Link:** 
[Link](https://www.typescriptlang.org/play/#src=declare%20const%20kind%3A%20unique%20symbol%0D%0A%0D%0Atype%20Square%20%3D%20%7B%0D%0A%20%20readonly%20%5Bkind%5D%3A%20%22square%22%0D%0A%20%20length%3A%20number%0D%0A%20%20width%3A%20number%0D%0A%7D%0D%0A%0D%0Atype%20Circle%20%3D%20%7B%0D%0A%20%20readonly%20%5Bkind%5D%3A%20%22circle%22%0D%0A%20%20radius%3A%20number%0D%0A%7D%0D%0A%0D%0Atype%20Shape%20%3D%20Square%20%7C%20Circle%0D%0A%0D%0Afunction%20doesNotNarrow(shape%3A%20Shape)%20%7B%0D%0A%20%20if%20(shape%5Bkind%5D%20%3D%3D%3D%20%22square%22)%20%7B%0D%0A%20%20%20%20console.log(shape.length%2C%20shape.width)%20%2F%2F%20error%20length%20does%20not%20exist%20on%20Shape%20%0D%0A%20%20%7D%20else%20%7B%0D%0A%20%20%20%20console.log(shape.radius)%20%2F%2F%20error%20radius%20does%20not%20exist%20on%20Shape%20%0D%0A%20%20%7D%0D%0A%7D)

**Related Issues:**

- #10530 _Discriminant property type guard not applied with bracket notation_ 

I know this looks like a straight duplicate but I don't think it should be considered one because symbols aren't mentioned anywhere in that issue or in the titles of any of its duplicates, and `unique symbol` did not exist when that issue was originally considered.

The only reason that #10530 hasn't been fixed is because it caused enough of a performance slowdown that it wasn't considered a worthwhile tradeoff. I'm hoping that limiting this to only `unique symbol`s will have a small enough performance impact to make it worthwhile this time.

Unlike most string keys, symbols can only be used with bracket notation so I think the value proposition here is much higher than it was for string keys. Plus it seems like using a symbol as a discriminant property that doesn't have to worry about conflicting with other property names is one of the most natural use cases for symbols.

**Comment by mhegazy** [comment]
> The only reason that #10530 hasn't been fixed is because it caused enough of a performance slowdown that it wasn't considered a worthwhile tradeoff. I'm hoping that limiting this to only unique symbols will have a small enough performance impact to make it worthwhile this time.

i am afraid the issue is not what you are indexing with. the issue is the binder currently marks places that can "possibly" narrow the type of a symbol, and puts them in a control flow graph. every time the type for a symbol is needed the graph is walked to get the type at that location. currently `a[..]` is not included as a possible narrowing location. adding it makes walking the control flow graph much more expensive (with all array accesses and such). It does not matter what it is that you are indexing with really.

**Comment by mhegazy** [comment]
We usually like to keep one ticket per issue. So closing in favor of #10530.

**Comment by mhegazy** [comment]
> Are you sure that this would be included if #10530 is ever fixed? 

I do not know. the issue is open still so i would not rule it out. I would recommend subscribing to the issue for further updates.

> If it isn’t likely to address that case then it seems possible that even #10530 wouldn’t fix symbols and this issue would still be relevant.

As i noted earlier, to allow narrowing on strings, numbers, symbols in element access expressions (`a[exp]`) we need to add that to the control flow graph, and that is what is causing the perf regression.
